internal/staging/debouncer_test.go

```diff
@@ -0,0 +1,72 @@
+package staging
+
+import (
+	"context"
+	"testing"
+	"time"
+)
+
+func TestDebouncer(t *testing.T) {
+	t.Run("callback is called only once during timeout period", func(t *testing.T) {
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		called := make(chan struct{}, 1)
+		debouncer := NewDebouncer(ctx, 100*time.Millisecond, func() {
+			called <- struct{}{}
+		})
+
+		go func() {
+			for range 5 {
+				debouncer.Debounce()
+				time.Sleep(20 * time.Millisecond)
+			}
+		}()
+
+		AssertCallbackCalled(t, called)
+
+		// Ensure no additional calls are made
+		AssertCallbackNotCalled(t, called)
+	})
+
+	t.Run("multiple callbacks called when requests sent over timeout", func(t *testing.T) {
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		called := make(chan struct{}, 1)
+		debouncer := NewDebouncer(ctx, 100*time.Millisecond, func() {
+			called <- struct{}{}
+		})
+
+		go func() {
+			for range 10 {
+				debouncer.Debounce()
+				time.Sleep(20 * time.Millisecond)
+			}
+		}()
+
+		AssertCallbackCalled(t, called)
+		AssertCallbackCalled(t, called)
+
+		// Ensure no additional calls are made
+		AssertCallbackNotCalled(t, called)
+	})
+}
+
+func AssertCallbackCalled(t *testing.T, called chan struct{}) {
```

Comment:
Nit: consider using `testify.mock` for this, and using eventually() for the assertions.